File: exercism/typescript/rational-numbers/rational-numbers.js
export function findGreatestCommonDivisor(a, b) {
    a = Math.abs(a);
    b = Math.abs(b);
    let gcd = 0;
    for (let i = 1; i <= a || i <= b; i++) {
        if (a % i === 0 && b % i === 0) {
            gcd = i;
        }
    }
    return gcd;
}
export class Rational {
    numerator;
    denominator;
    constructor(a, b) {
        this.numerator = a;
        if (b !== 0) {
            this.denominator = b;
        }
        else {
            throw new Error("Denominator cannot be 0");
        }
    }
    add(other) {
        let a1 = this.numerator;
        let b1 = this.denominator;
        let a2 = other.numerator;
        let b2 = other.denominator;
        this.numerator = a1 * b2 + a2 * b1;
        this.denominator = b1 * b2;
        return this.reduce();
    }
    sub(other) {
        let a1 = this.numerator;
        let b1 = this.denominator;
        let a2 = other.numerator;
        let b2 = other.denominator;
        this.numerator = a1 * b2 - a2 * b1;
        this.denominator = b1 * b2;
        return this.reduce();
    }
    mul(other) {
        let a1 = this.numerator;
        let b1 = this.denominator;
        let a2 = other.numerator;
        let b2 = other.denominator;
        this.numerator = a1 * a2;
        this.denominator = b1 * b2;
        return this.reduce();
    }
    div(other) {
        let a1 = this.numerator;
        let b1 = this.denominator;
        let a2 = other.numerator;
        let b2 = other.denominator;
        this.numerator = a1 * b2;
        this.denominator = a2 * b1;
        return this.reduce();
    }
    abs() {
        this.numerator = Math.abs(this.numerator);
        this.denominator = Math.abs(this.denominator);
        return this.reduce();
    }
    exprational(exponential) {
        this.numerator = this.numerator ** exponential;
        this.denominator = this.denominator ** exponential;
        return this.reduce();
    }
    expreal(exponential) {
        // lack of precision 8^(4/3) = ~16 for Math.pow(8, 4/3); Math.cbrt returned more accurate result
        if (this.denominator === 3) {
            return Math.cbrt(exponential ** this.numerator);
        }
        return Math.pow(exponential, this.numerator / this.denominator);
    }
    reduce() {
        if (this.numerator === 0) {
            this.denominator = 1;
            return this;
        }
        let gcd = findGreatestCommonDivisor(this.numerator, this.denominator);
        this.numerator /= gcd;
        this.denominator /= gcd;
        if ((this.numerator < 0 && this.denominator < 0) || (this.numerator > 0 && this.denominator < 0)) {
            this.numerator = -this.numerator;
            this.denominator = -this.denominator;
        }
        return this;
    }
}

File: exercism/python/rational-numbers/rational_numbers.py
"""Rational number implementation with normalized arithmetic operations."""

from __future__ import annotations

import math


class Rational:
    """Represent a reduced rational number in standard sign form."""

    def __init__(self, numer: int, denom: int) -> None:
        if not denom:
            raise ZeroDivisionError("Denominator cannot be zero.")

        greatest_common_divisor = math.gcd(numer, denom)
        reduced_numer = numer // greatest_common_divisor

        if (reduced_denom := denom // greatest_common_divisor) < 0:
            reduced_numer *= -1
            reduced_denom *= -1

        if not reduced_numer:
            reduced_denom = 1

        self.numer = reduced_numer
        self.denom = reduced_denom

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Rational):
            return NotImplemented
        return self.numer == other.numer and self.denom == other.denom

    def __hash__(self) -> int:
        return hash((self.numer, self.denom))

    def __repr__(self) -> str:
        return f"{self.numer}/{self.denom}"

    def __add__(self, other: Rational) -> Rational:
        result_numer = self.numer * other.denom + other.numer * self.denom
        result_denom = self.denom * other.denom
        return Rational(result_numer, result_denom)

    def __sub__(self, other: Rational) -> Rational:
        result_numer = self.numer * other.denom - other.numer * self.denom
        result_denom = self.denom * other.denom
        return Rational(result_numer, result_denom)

    def __mul__(self, other: Rational) -> Rational:
        result_numer = self.numer * other.numer
        result_denom = self.denom * other.denom
        return Rational(result_numer, result_denom)

    def __truediv__(self, other: Rational) -> Rational:
        result_numer = self.numer * other.denom
        result_denom = self.denom * other.numer
        return Rational(result_numer, result_denom)

    def __abs__(self) -> Rational:
        return Rational(abs(self.numer), abs(self.denom))

    def __pow__(self, power: int) -> Rational:
        if power >= 0:
            return Rational(self.numer**power, self.denom**power)

        absolute_power = abs(power)
        return Rational(self.denom**absolute_power, self.numer**absolute_power)

    def __rpow__(self, base: float) -> float:
        exponent_value = self.numer / self.denom
        return float(base**exponent_value)
